### PR TITLE
IPR / Attribution

### PR DESCRIPTION
**Pull request by vcoindet**
IPR Attribution

In ApiGlobe.js

getLayersIPR() -> return an array of IPR with no duplicate
setLayerIPR(param)
removeDuplicateFomArray(originalArray, prop)

Note that the name of the IPR is mandatory

### CODE REVIEW

**Comment by vcoindet** [comment]
rebased and ready

**Comment by vcoindet** [comment]
Have to do some changes

**Comment by vcoindet** [comment]
Ready

File: index.html [hunk 1]
```diff
@@ -55,6 +55,8 @@
             urlImageryLayers.push('examples/layers/JSONLayers/OrthosCRS.json');
             urlImageryLayers.push('examples/layers/JSONLayers/ScanEX.json');
             urlImageryLayers.push('examples/layers/JSONLayers/Region.json');
+            /* urlImageryLayers.push('examples/layers/JSONLayers/DARK.json');
+            urlImageryLayers.push('examples/layers/JSONLayers/OPENSM.json'); */
```
Comment (autra) [inline index.html]: Maybe remove these comments?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 2]
```diff
@@ -852,6 +852,62 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
     this.viewerDiv.removeEventListener('layerchanged:index', this.callbackLayerChanged, false);
 };
 
+/**
+ * Get the IPR (Intellectual Property Rights) of all layers in the scene.
+ * @constructor
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Not a constructor.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 3]
```diff
@@ -852,6 +852,62 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+ * @return {array}  An array of IPR.
+ */
+
+ApiGlobe.prototype.getLayersIPR = function getLayersIPR() {
+    var IPR = [];
+    var colorLayers = this.scene.getMap().layersConfiguration.getColorLayers();
+    for (var i = 0; i < colorLayers.length; i++) {
+        IPR.push(colorLayers[i].options.ipr);
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: You could use `Array.prototype.map` (or at least a `for...of` loop).

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 4]
```diff
@@ -852,6 +852,62 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+    }
+    var elevationLayers = this.scene.getMap().layersConfiguration.getElevationLayers();
+    for (var j = 0; j < elevationLayers.length; j++) {
+        IPR.push(elevationLayers[j].options.ipr);
+    }
+    IPR = this.removeDuplicatesFromArray(IPR, 'name');
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Why not using a Set?

Comment (autra) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: It would avoid to implement another function just for that

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 5]
```diff
@@ -852,6 +852,62 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
 
+/**
+ * Get the IPR (Intellectual Property Rights) of all layers in the scene.
+ * @constructor
+ * @return {array}  An array of IPR.
+ */
+
+ApiGlobe.prototype.getLayersIPR = function getLayersIPR() {
+    var IPR = [];
+    var colorLayers = this.scene.getMap().layersConfiguration.getColorLayers();
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Use `const` for both vars.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 6]
```diff
@@ -852,6 +852,62 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+    }
+    for (i in lookupObject) {
+        newArray.push(lookupObject[i]);
+    }
+    return newArray;
+};
+
+/**
+ * Set the IPR for a layer.
+ * @constructor
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: not a constructor

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 7]
```diff
@@ -852,6 +852,62 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+ * @param {param} Param - The id of the layer and the new IPR {id, ipr}.
+ */
+
+ApiGlobe.prototype.setLayerIPR = function setLayerIPR(param) {
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Do we really need this method if it is specified in the json? These are mainly static info anyway right?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 8]
```diff
@@ -852,6 +852,62 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+    if (param.ipr.name === undefined) {
+        throw new Error(`Missing parameter "param.ipr.name" : ${param.id}`);
+    }
+    var colorLayers = this.scene.getMap().layersConfiguration.getColorLayers();
+    for (var i = 0; i < colorLayers.length; i++) {
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: For..of loops

File: src/Core/Commander/Providers/WMS_Provider.js [hunk 9]
```diff
@@ -67,12 +67,20 @@ WMS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer)
     }
 
     layer.format = layer.options.mimetype || 'image/png';
+    layer.options.ipr = layer.options.ipr || '';
+    layer.options.ipr.description = layer.options.ipr.description || '';
+    layer.options.ipr.url = layer.options.ipr.url || '';
+    layer.options.ipr.logo = layer.options.ipr.logo || '';
+    layer.options.ipr.name = layer.options.ipr.name;
```
Comment (autra) [inline src/Core/Commander/Providers/WMS_Provider.js]: well, I guess we don't need this line :-)

File: src/Core/Commander/Providers/WMS_Provider.js [hunk 10]
```diff
@@ -67,12 +67,20 @@ WMS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer)
...
     layer.crs = layer.projection || 'EPSG:4326';
     layer.width = layer.heightMapWidth || 256;
     layer.version = layer.version || '1.3.0';
     layer.style = layer.style || '';
     layer.transparent = layer.transparent || false;
     layer.bbox = layer.bbox || new BoundingBox();
+    if (layer.options.ipr.name === undefined) {
+        throw new Error(`Missing parameter "layer.options.ipr.name" in layer : ${layer.id}`);
```
Comment (autra) [inline src/Core/Commander/Providers/WMS_Provider.js]: Is this really a requirement? Is it possible for a data source to *not* have any known Intellectual Property?

File: src/Core/Commander/Providers/WMTS_Provider.js [hunk 11]
```diff
@@ -71,6 +71,13 @@ WMTS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer
         options.mimetype = options.mimetype || 'image/png';
         options.style = options.style || 'normal';
         options.projection = options.projection || 'EPSG:3857';
+        options.ipr.description = options.ipr.description || '';
+        options.ipr.url = options.ipr.url || '';
+        options.ipr.logo = options.ipr.logo || '';
+        options.ipr.name = options.ipr.name;
```
Comment (autra) [inline src/Core/Commander/Providers/WMTS_Provider.js]: So all this defensive code will be virtually identical for every provider. Why not moving it in `ApiGlobe.js:preprocessLayer`? This method has already some similar initialization code and it would avoid duplicate codes.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 12]
```diff
@@ -852,6 +852,62 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+    }
+    var elevationLayers = this.scene.getMap().layersConfiguration.getElevationLayers();
+    for (var j = 0; j < elevationLayers.length; j++) {
+        IPR.push(elevationLayers[j].options.ipr);
+    }
+    IPR = this.removeDuplicatesFromArray(IPR, 'name');
+    return IPR;
+};
+
+ApiGlobe.prototype.removeDuplicatesFromArray = function removeDuplicatesFromArray(originalArray, prop) {
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: This should not be part of the api. It should be in a `utils` module somewhere. Or at least not in the prototype. But I actually think you don't need this, see my above comments.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 13]
```diff
@@ -852,6 +852,62 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+    for (var i = 0; i < colorLayers.length; i++) {
+        IPR.push(colorLayers[i].options.ipr);
+    }
+    var elevationLayers = this.scene.getMap().layersConfiguration.getElevationLayers();
+    for (var j = 0; j < elevationLayers.length; j++) {
+        IPR.push(elevationLayers[j].options.ipr);
+    }
+    IPR = this.removeDuplicatesFromArray(IPR, 'name');
+    return IPR;
+};
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: So to summarize, another possible implementation:

```js 
ApiGlobe.prototype.getLayersIPR = function getLayersIPR() {
    const lc = this.scene.getMap().layersConfiguration; // just a shortcut
    const result = new Set();
   // the spread operator allows for quick concatenation 
    return [...lc.getColorLayers(), ...lc.getElevationLayers()].forEach(l => result.add(l.options.ipr));
   // or return new Array(result) if you really need an array.
   // A Set seems more suitable here to me here, though.
};
```

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 7 above)
Comment (autra) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: I mean, we're not supposed to change them during the execution right?

**Review by autra — COMMENTED** [review]
Hey \o 

Just passing by, left some comments where I think things can be fancier :-)

File: index.html (hunk 1 above)
Comment (vcoindet) [reply index.html]: Yes, I forgot to remove them (I used them for my tests)

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 7 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Well, it is required by the IGN's geoportal

File: src/Core/Commander/Providers/WMS_Provider.js (hunk 10 above)
Comment (vcoindet) [reply src/Core/Commander/Providers/WMS_Provider.js]: I will check that, I will discuss this with the IGN's Geoportal

File: src/Core/Commander/Providers/WMTS_Provider.js (hunk 11 above)
Comment (vcoindet) [reply src/Core/Commander/Providers/WMTS_Provider.js]: OK, I will check that

File: index.html (hunk 1 above)
Comment (vcoindet) [reply index.html]: done

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 2 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: done

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 13 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Wasn't exactly what I wanted, I used a Map instead.
I also used `for...of` loop

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 12 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: removed

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 6 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: done

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 7 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Some layers might not have any IPR specified in the json, so it's interesting to have a method to set an IPR for those layers.

File: src/Core/Commander/Providers/WMS_Provider.js (hunk 9 above)
Comment (vcoindet) [reply src/Core/Commander/Providers/WMS_Provider.js]: removed

File: src/Core/Commander/Providers/WMS_Provider.js (hunk 10 above)
Comment (vcoindet) [reply src/Core/Commander/Providers/WMS_Provider.js]: removed

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 7 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Note that I specified the attribut `param.ipr.name` mandatory in this method. If one choose to specify an IPR, this one must have one argument at least.

**Comment by vcoindet** [comment]
rebased and ready

File: examples/layers/JSONLayers/DARK.json [hunk 14]
```diff
@@ -3,6 +3,12 @@
     "id": "DARK",
     "customUrl": "http://a.basemaps.cartocdn.com/dark_all/%TILEMATRIX/%COL/%ROW.png",
     "options": {
+    	"ipr": {
+    		"name":"CARTO",
+    		"url": "https://carto.com/",
+    		"logo":"",
```
Comment (peppsac) [inline examples/layers/JSONLayers/DARK.json]: What's the point of adding empty properties (logo, description)?

File: src/Core/Commander/Providers/WMS_Provider.js [hunk 15]
```diff
@@ -73,7 +73,6 @@ WMS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer)
     layer.style = layer.style || '';
     layer.transparent = layer.transparent || false;
     layer.bbox = layer.bbox || new BoundingBox();
-
```
Comment (peppsac) [inline src/Core/Commander/Providers/WMS_Provider.js]: Unrelated change (it adds noise to the PR so I'd rather not have this change here)

**Review by peppsac — CHANGES_REQUESTED** [review]

**Comment by autra** [comment]
Actually... @vcoindet can you explain to me the rationale of adding such apis in Itowns? I actually don't see why itowns should be aware of such a thing. It is the client application (here the geoportal) which adds layers to itowns (which has the ambition of being a framework, let's recall that), so this geoportal is perfectly aware of these IPR informations and can display them independently (they are not part of WMTS nor WMS specs right?). It can even query the servers independently of itowns.

I really feel that the reasoning is backward here. Itowns should not be (in my humble opinion) a helper for the geoportal. Rather, it should focus on it's goal : be able to display all the layers (of all the possible type) as easily as possible, and allow to configure stuff like cameras, controls etc. Thoughts?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 16]
```diff
@@ -852,6 +852,50 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+ * @return {array}  An array of IPR.
+ */
+
+ApiGlobe.prototype.getLayersIPR = function getLayersIPR() {
+    const lc = this.scene.getMap().layersConfiguration;
+    const colorLayers = lc.getColorLayers();
+    const map = new Map();
+    for (const i in colorLayers) {
+        map.set(colorLayers[i].options.ipr.name, colorLayers[i].options.ipr);
+    }
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: I meant 

```js 
for (const layer of lc.getColorLayers()) {
    map.set(layer.options.ipr.name, layer.options.ipr);
}
```
- use for..of loops
- Matter of personal opinion, but no need to have a variable `colorLayers` when you are accessing the field only once.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 17]
```diff
@@ -852,6 +852,50 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+    const elevationLayers = lc.getElevationLayers();
+    for (const j in elevationLayers) {
+        map.set(elevationLayers[j].options.ipr.name, elevationLayers[j].options.ipr);
+    }
+    map.delete(undefined);
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: ?

**Comment by nosy-b** [comment]
@autra I share your point of view but adding an optional credit properties in the layer description will not transform iTowns into a heavy library.
Yes, French Geoportal or other website using itowns should then handle themself how they want the credit to be displayed.

**Comment by vcoindet** [comment]
Well, I guess that itowns should be able to display all the layers, but ain't we supposed to mention their sources ?
I agree that some layers won't specify their IPR in their files, that's why I removed the defensive code WMS_Provider et WMTS_Provider.

It won't be useful for the geoportal only, as one is supposed to mention the sources of the layers displayed on iTowns... That's why I developed these methods.

**Comment by gcebelieu** [comment]
@autra , @nosy-b  I agree with @vcoindet answer : IPR feature goes further than a simple geoportal need : you may have to quote geographical data originators to meet their conditions of use when they exist. 

If this may reassure you : this feature is proposed by many frameworks such as CesiumJS, OpenLayers or Leaflet...

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 16 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: ok, I see

**Comment by autra** [comment]
> If this may reassure you : this feature is proposed by many frameworks such as CesiumJS, OpenLayers or Leaflet...

Because they display it. If we want to display it too, then I suggest adding the displaying code to this PR. It's a logically linked change and would make the rationale more obvious :-) If we do not, then really we don't need this API imho.

Also, let's call the object `attribution` and not `IPR`, like Cesium, Leaflet etc... all do. `attribution` is now apparently a de facto standard and it would make more sense for newcomers who are used to these frameworks.

Furthermore, another reason why it's important to have the displaying code: it affects the api design. For example at the moment you're returning a map. Whether or not this is relevant will depend on how we use this API. FYI, Cesium, Leaflet and OpenLayers all gives back html and not a Map. 

Concerning the IPR internals, they use a simpler model: Cesium and Leaflet have only a text field (a description). OpenLayers adds an href. Do we really need a logo and 2 text fields? Maybe description and href are enough (and indeed you did fill `description` and `logo` to empty strings in your PR), WDYT?

I guess you still need a defensive code in preprocessLayer, as attributions are optionals: they might not be set in the json.

**Comment by autra** [comment]
Also not sure `ApiGlobe.js` is the right place for that. There should be some `layer` objects like OpenLayers has where you can add these attributions, but I guess rationalizing the api is not the goal of this PR...

File: examples/layers/JSONLayers/DARK.json (hunk 14 above)
Comment (vcoindet) [reply examples/layers/JSONLayers/DARK.json]: removed

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 16 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: done

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 17 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Some IPR might not be specified, so I chose to remove it from the map object

**Comment by vcoindet** [comment]
Rebased and Ready

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 18]
```diff
@@ -917,6 +917,39 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
     this.viewerDiv.removeEventListener('layerchanged:index', this.callbackLayerChanged, false);
 };
 
+/**
+ * Get the IPR (Intellectual Property Rights) of all layers in the scene.
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Let's update the comments too

File: examples/layers/JSONLayers/DARK.json [hunk 19]
```diff
@@ -3,6 +3,10 @@
     "id": "DARK",
     "customUrl": "http://a.basemaps.cartocdn.com/dark_all/%TILEMATRIX/%COL/%ROW.png",
     "options": {
+    	"ipr": {
```
Comment (autra) [inline examples/layers/JSONLayers/DARK.json]: let's call them "attribution" ?

File: src/Core/Commander/Providers/WMS_Provider.js (hunk 15 above)
Comment (autra) [reply src/Core/Commander/Providers/WMS_Provider.js]: +1, would pollute `git log -- src/Core/Commander/Providers` for example.

**Comment by vcoindet** [comment]
rebased and ready

**Comment by autra** [comment]
IIRC after discussing with the geoportal team, we just need to expose an event containing `zoom` and `extent` changes, then they would calculate which attribution to display (according to their metadata) themselves. So I guess we don't need this PR any more right?

**Comment by gcebelieu** [comment]
> So I guess we don't need this PR any more right?

well, there is still a need that can be reusable for itowns framework. Let me expose it :

I think iTowns API should give the possibility to attach one or several Attribution(s) object(s)  to a layer object (the object structure may be discussed further). 

The Attribution Widget role is to aggregate all attributions attached to layers currently visible on the globe (this aggregation is processed at every layer change event that occurs on the globe) and display an aggregated result on a given DOM target. The aggregation may be smart enough to avoid displaying the same attribution twice.

The final application (here geoportail) will listen to zoom and center changed events in order to dynamicaly feed Attribution property of displayed layers. The Attribution Widget will then refresh itself automaticaly.

What do you think ?

**Comment by autra** [comment]
> I think iTowns API should give the possibility to attach one or several Attribution(s) object(s) to a layer object (the object structure may be discussed further).

Why not? but:

> The Attribution Widget role is to aggregate all attributions attached to layers currently visible on the globe (this aggregation is processed at every layer change event that occurs on the globe) and display an aggregated result on a given DOM target. The aggregation may be smart enough to avoid displaying the same attribution twice.

So not a widget, just a div on one corner of Itowns, like Leaflet does for instance. In this case, I repeat myself: this PR should include the displaying code.

> The final application (here geoportail) will listen to zoom and center changed events in order to dynamicaly feed Attribution property of displayed layers. The Attribution Widget will then refresh itself automaticaly.

So what you are suggesting is 
construct time
1/ itowns provides a template file (Widget.js) 
2/ Geoportail instanciate a class MyWidget.js
3/ geoportail feeds this instance to itowns
4/ itowns remembers it

then runtime !
5/ user zoom
6/ itowns sends an event
7/ geoportal code (in MyWidget probably) get the event and extract a metadata
8/ geoportal sends the metadata *back* to itowns
9/ itowns orders the widget to refresh, calling a function on it.

Don't you see how this is a nonsense technically? Especially steps 7, 8 and 9 ? What I'm saying is simple: you don't need itowns to manage your widgets, and you'll be better off. Free yourself :-) Really, I'm not speaking only as a itowns dev. I'm also a user of itowns on other projects. Forcing myself to implement (nonsense in javascript) a "class" whose structure is forced on me by the framework is a pain.

Beside it is: 
1/ Orthogonal to the current PR (most of all, you want zoom/center changed events).
2/ highly specific to geoportail needs, and imho this dynamic behaviour should not be in Itowns core. It's currently not clear and debatable that this need would be shared by other client apps.

Again, you just need to have a `ExtendChangedEvent`, and your client code would just be

```js
itowns.viewer.addEventListener('extend-changed-event', function(ev) {
    let extend = ev.extend;
    let zoom = ev.zoom
    /* your code in step 7 */
   /* your code in function called during step 9 */
});
```
Don't you see how it is *simpler*, for everybody?

**Comment by autra** [comment]
That being said 

> highly specific to geoportail needs, and imho this dynamic behaviour should not be in Itowns core. 

is my opinion. I guess the best course of action for this feature request would be to open an issue tagged with `enhancement`.

**Comment by gcebelieu** [comment]
> So not a widget, just a div on one corner of Itowns, like Leaflet does for instance. In this case, I repeat myself: this PR should include the displaying code.

Yeah, I totaly agree with that : this widget (or whatever you want to call it) should also display its result in a div.

> So what you are suggesting is
> construct time
> ... 
> Don't you see how this is a nonsense technically? Especially steps 7, 8 and 9 ? What I'm saying is simple: you don't need itowns to manage your widgets, and you'll be better off. Free yourself :-) Really, I'm not speaking only as a itowns dev. I'm also a user of itowns on other projects. Forcing myself to implement (nonsense in javascript) a "class" whose structure is forced on me by the framework is a pain.


I was not writting about a generic Widget (we've already discussed and agreed about that : "let's start with events and see" ) : I was just trying to explain what could be the perimeter of that attribution widget in iTowns API and how Geoportail (which has more complex attributions management) may handle with it (and event).

> I guess the best course of action for this feature request would be to open an issue tagged with enhancement

This issue has been created almost a year ago (#60 with an agreement of @vpicavet ) and conducted to this PR. We certainely may refine it and discuss about the perimeter, but please, let's go on forward on that subject.

**Comment by vcoindet** [comment]
Rebased and Ready

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 20]
```diff
@@ -917,6 +917,39 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+/**
+ * Get the Attribution of all layers in the scene.
+ * @return {array}  An array of attribution.
+ */
+
+ApiGlobe.prototype.getLayersAttribution = function getLayersAttribution() {
+    const lc = this.scene.getMap().layersConfiguration;
+    const map = new Map();
+    [...lc.getColorLayers(), ...lc.getElevationLayers()].forEach((l) => {
+        if (l.options.attribution != undefined) {
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: if (l.options.attribution != undefined)

replace by

if (!l.options.attribution)

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 21]
```diff
@@ -917,6 +917,39 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
     this.viewerDiv.removeEventListener('layerchanged:index', this.callbackLayerChanged, false);
 };
 
+/**
+ * Get the Attribution of all layers in the scene.
+ * @return {array}  An array of attribution.
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Return map not Array

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 22]
```diff
@@ -917,6 +917,39 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+    return map;
+};
+
+/**
+ * Set the attribution for a layer.
+ * @param {param} Param - The id of the layer and the new IPR {id, attribution}.
+ */
+
+ApiGlobe.prototype.setLayerAttribution = function setLayerAttribution(param) {
+    if (param.attribution.name === undefined) {
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: if (!param.attribution.name)

**Review by gchoqueux — CHANGES_REQUESTED** [review]

**Comment by gchoqueux** [comment]
I rebase this branch.

@peppsac Have you seen the changes?

**Comment by vcoindet** [comment]
changes done @gchoqueux  and @peppsac 
rebased and ready

**Comment by vcoindet** [comment]
rebased

**Comment by vcoindet** [comment]
rebased

**Comment by vcoindet** [comment]
rebased and ready @gchoqueux and @peppsac

**Review by gchoqueux — APPROVED** [review]

**Comment by vcoindet** [comment]
If no one has any further change requested, I propose that this PR be merged by wenesday the 15th.
@peppsac @gchoqueux WDYT ?

File: src/Core/Commander/Providers/WMS_Provider.js [hunk 23]
```diff
@@ -69,6 +69,7 @@ WMS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer)
 
     layer.bbox_url = layer.bbox_url || 'swne';
     layer.format = layer.options.mimetype || 'image/png';
+    layer.crs = layer.projection || 'EPSG:4326';
```
Comment (peppsac) [inline src/Core/Commander/Providers/WMS_Provider.js]: Unrelated change, please remove.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 24]
```diff
@@ -915,6 +915,39 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+    });
+    return map;
+};
+
+/**
+ * Set the attribution for a layer.
+ * @param {param} Param - The id of the layer and the new IPR {id, attribution}.
+ */
+
+ApiGlobe.prototype.setLayerAttribution = function setLayerAttribution(param) {
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: I'm not sure this function is useful at all.
Is there a use case for modifying attribution at runtime ?
And also :
```js
api.setLayerAttribution({id: layer.id, attribution: { ...});
```
can be written in a simpler way:
```js
layer.attribution = { ... }
```
WDYT?

File: src/Core/Commander/Providers/WMS_Provider.js (hunk 23 above)
Comment (vcoindet) [reply src/Core/Commander/Providers/WMS_Provider.js]: Removed

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 24 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: I agree with you, it's simplier to do that way, I removed the function.

**Comment by vcoindet** [comment]
Rebased and ready @peppsac

**Review by peppsac — APPROVED** [review]

**Comment by peppsac** [comment]
LGTM, merging.
